### PR TITLE
WIP: Autogen speedups

### PR DESCRIPTION
**Pull request by jvilk-stripe**
I'm opening this WIP PR to test some autogen speedups. There's nothing to see here right now; once I iterate on testing these a bit I'll write up my findings.

### Motivation

Speed up autogen.

### Test plan

Covered by existing tests (hopefully?).

### CODE REVIEW

**Comment by jvilk-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JjQ5Eh8yORQdg0
→ https://go/builds/bui_JjQ5VKDUEuiakG

**Comment by jvilk-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JjQKJ4LhB8yBnI
→ https://go/builds/bui_JjQKwhOMOmuLVJ

**Comment by jvilk-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JjktgccRhOxusZ
→ https://go/builds/bui_JjktUDXRDk3VxW

**Comment by jvilk-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_Jjkz4JemhYhAsb
→ https://go/builds/bui_JjkzZoekOOGDIH

**Comment by jvilk-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_Jjl8sUddwSGFDU
→ https://go/builds/bui_Jjl85evHYDEWnB

**Comment by jvilk-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JjnoyQXRwbaDAj
→ https://go/builds/bui_Jjnp30b7F55dRi

File: main/autogen/autogen.cc [hunk 1]
```diff
@@ -326,13 +325,13 @@ class AutogenWalk {
...
 // the current file
-ParsedFile Autogen::generate(core::Context ctx, ast::ParsedFile tree) {
+ParsedFile Autogen::generate(core::Context ctx, ast::ParsedFile tree, const CRCBuilder &crcBuilder) {
     AutogenWalk walk;
     tree.tree = ast::TreeMap::apply(ctx, walk, move(tree.tree));
     auto pf = walk.parsedFile();
     pf.path = string(tree.file.data(ctx).path());
     auto src = tree.file.data(ctx).source();
-    pf.cksum = CRC::Calculate(src.data(), src.size(), CRC::CRC_32());
+    pf.cksum = crcBuilder.crc32(src);
```
Comment (froydnj) [inline main/autogen/autogen.cc:336]: Heh, nice find.  Is crc computation hot enough that we could consider using the crc instructions (SSE4.2) on x86-64?

Comment (jvilk-stripe) [reply main/autogen/autogen.cc:336]: It's ~30% of this function's (Autogen::generate) runtime according to perf, so yeah! Do we have that instruction on devboxen?

Comment (froydnj) [reply main/autogen/autogen.cc:336]: We do!  https://github.com/komrad36/CRC#option-11-1-byte-hardware-accelerated and following has a nice (if long) explanation of how to best use the instruction.  (Search for "OPTION 13" if you just want the gofast version and don't want any explanation of what's going on.)

**Comment by jvilk-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_Jkt4xAnp0FUf8V
→ https://go/builds/bui_Jkt4p5nUEUqOd3

File: main/autogen/autogen.cc (hunk 1 above)
Comment (jvilk-stripe) [reply main/autogen/autogen.cc:336]: That's great. Seems like we could expect a ~10X speedup.

Comment (jvilk-stripe) [reply main/autogen/autogen.cc:336]: Looks like we assume at least sandybridge, so this could be the default CRC calculation method w/ no fallback.

**Comment by jvilk-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_Jktf9KMLi9Is2z
→ https://go/builds/bui_JktfY97L3DZHus

**Comment by jvilk-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JkuDGjO40Fj4fx
→ https://go/builds/bui_JkuD9WRGKv7l9I

**Comment by jvilk-stripe** [comment]
I've moved out all of the optimizations in this PR into other PRs.

**Pull request closed without merging**